src/mo_idl/mo_to_idl.ml

```diff
@@ -44,74 +48,86 @@ module MakeState() = struct
              (k, n + 1))
-        | Some (k, n) -> (k, n)
+        | Some kn -> kn
       in
       begin
         match (k, n) with
+        | _ when k < 0 || n < 0 -> assert false
         | (0, 0) -> name
-        | (0, n) when n > 0 -> Printf.sprintf "%s_%d" name n
-        | (k, 0) when k > 0 -> Printf.sprintf "%s__%d" name k
-        | (k, n) when k > 0 && n > 0 -> Printf.sprintf "%s__%d_%d" name k n
-        | _ -> assert false
+        | (0, n) -> Printf.sprintf "%s_%d" name n
+        | (k, 0) -> Printf.sprintf "%s__%d" name k
+        | (k, n) -> Printf.sprintf "%s__%d_%d" name k n
       end
     | _ -> assert false
 
-  let prim = function
-    | Null -> I.PrimT I.Null
-    | Bool -> I.PrimT I.Bool
-    | Nat -> I.PrimT I.Nat
-    | Nat8 -> I.PrimT I.Nat8
-    | Nat16 -> I.PrimT I.Nat16
-    | Nat32 -> I.PrimT I.Nat32
-    | Nat64 -> I.PrimT I.Nat64
-    | Int -> I.PrimT I.Int
-    | Int8 -> I.PrimT I.Int8
-    | Int16 -> I.PrimT I.Int16
-    | Int32 -> I.PrimT I.Int32
-    | Int64 -> I.PrimT I.Int64
-    | Float -> I.PrimT I.Float64
-    | Char -> I.PrimT I.Nat32
-    | Text -> I.PrimT I.Text
-    | Blob -> I.BlobT
-    | Principal -> I.PrincipalT
+  let prim = let open I in
+    function
+    | Type.Null -> PrimT Null
+    | Bool -> PrimT Bool
+    | Nat -> PrimT Nat
+    | Nat8 -> PrimT Nat8
+    | Nat16 -> PrimT Nat16
+    | Nat32 -> PrimT Nat32
+    | Nat64 -> PrimT Nat64
+    | Int -> PrimT Int
+    | Int8 -> PrimT Int8
+    | Int16 -> PrimT Int16
+    | Int32 -> PrimT Int32
+    | Int64 -> PrimT Int64
+    | Float -> PrimT Float64
+    | Char -> PrimT Nat32
+    | Text -> PrimT Text
+    | Blob -> BlobT
+    | Principal -> PrincipalT
     | Region
     | Error -> assert false
 
   let rec typ t =
     (match t with
-    | Any -> I.PrimT I.Reserved
-    | Non -> I.PrimT I.Empty
+    | Any -> I.(PrimT Reserved)
+    | Non -> I.(PrimT Empty)
     | Prim p -> prim p
     | Var (s, i) -> assert false
     | Con (c, ts) ->
        (match Cons.kind c with
         | Def (_, t) ->
            (match (open_ ts t) with
             | Prim p -> prim p
-            | Any -> I.PrimT I.Reserved
-            | Non -> I.PrimT I.Empty
+            | Any -> I.(PrimT Reserved)
+            | Non -> I.(PrimT Empty)
             | t ->
-               let id = monomorphize_con ts c in
-               if not (Env.mem id !env) then
-                 begin
-                   env := Env.add id (I.PreT @@ no_region) !env;
-                   let t = typ (normalize t) in
-                   env := Env.add id t !env
-                 end;
-               I.VarT (id @@ no_region))
+              let id = monomorphize_con ts c in
+              if Env.mem id !env then
+                match Env.find id !env with
+                | I.PreT -> I.VarT (id @@ no_region)
+                | I.VarT _ as seen -> seen
+                | seen -> I.VarT (RevMap.find (Cons.name c, seen) !rev @@ no_region)
+              else begin
```

Comment:
```suggestion
              match Env.find_opt id !env with
              | Some t -> ( match t with
                | I.PreT -> I.VarT (id @@ no_region)
                | I.VarT _ as seen -> seen
                | seen -> I.VarT (RevMap.find (Cons.name c, seen) !rev @@ no_region))
             | None ->
```

Reply:
Yeah, further removed the nested `match`: 385b419f682a39547099189c6666f85721088b8b